bundle/python/transform.go

```diff
@@ -0,0 +1,106 @@
+package python
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/bundle/artifacts"
+	"github.com/databricks/cli/bundle/libraries"
+	"github.com/databricks/databricks-sdk-go/service/compute"
+	"github.com/databricks/databricks-sdk-go/service/jobs"
+)
+
+// This mutator takes the wheel task and trasnforms it into notebook
+// which installs uploaded wheels using %pip and then calling corresponding
+// entry point.
+func TransformWheelTask() bundle.Mutator {
+	return &transform{}
+}
+
+type transform struct {
+}
+
+func (m *transform) Name() string {
+	return "python.TransformWheelTask"
+}
+
+const INSTALL_WHEEL_CODE = `%%pip install --force-reinstall %s`
+
+const NOTEBOOK_CODE = `
+%%python
+%s
+
+from contextlib import redirect_stdout
+import io
+import sys
+sys.argv = [%s]
+
+import pkg_resources
+_func = pkg_resources.load_entry_point("%s", "console_scripts", "%s")
+
+f = io.StringIO()
+with redirect_stdout(f):
+	_func()
+s = f.getvalue()
+dbutils.notebook.exit(s)
+`
+
+func (m *transform) Apply(ctx context.Context, b *bundle.Bundle) error {
+	// TODO: do the transformaton only for DBR < 13.1 and (maybe?) existing clusters
+	wheelTasks := libraries.FindAllWheelTasks(b)
+	for _, wheelTask := range wheelTasks {
+		taskDefinition := wheelTask.PythonWheelTask
+		libraries := wheelTask.Libraries
+
+		wheelTask.PythonWheelTask = nil
+		wheelTask.Libraries = nil
+
+		path, err := generateNotebookWrapper(taskDefinition, libraries)
+		if err != nil {
+			return err
+		}
+
+		remotePath, err := artifacts.UploadNotebook(context.Background(), path, b)
+		if err != nil {
+			return err
+		}
+
+		os.Remove(path)
+
+		wheelTask.NotebookTask = &jobs.NotebookTask{
+			NotebookPath: remotePath,
+		}
+	}
+	return nil
+}
+
+func generateNotebookWrapper(task *jobs.PythonWheelTask, libraries []compute.Library) (string, error) {
+	pipInstall := ""
+	for _, lib := range libraries {
+		pipInstall = pipInstall + "\n" + fmt.Sprintf(INSTALL_WHEEL_CODE, lib.Whl)
+	}
+	content := fmt.Sprintf(NOTEBOOK_CODE, pipInstall, generateParameters(task), task.PackageName, task.EntryPoint)
```

Comment:
Instead of `sprintf` we should use `text/template` to make this a bit more extensible, IMO.

The output of this generator is always going to be a `[]NotebookCell`. For Databricks notebooks, we can include the text verbatim (joined on `\n`) but for ipynb notebooks we'll have to do something smarter and prepend the cells in the JSON.